library/general/src/lib/cfa/nsswitch.rb

```diff
@@ -0,0 +1,273 @@
+    def save
+      return false unless modified?
+
+      new_content = @parser.serialize(data)
+
+      @parser.file_name = write_path if @parser.respond_to?(:file_name=)
+      @file_handler.write(write_path, new_content)
+      @current_content = new_content
+
+      true
+    end
+
+    # Whether the content has changed
+    #
+    # @return [Boolean] true if content has been modified; false otherwise.
+    def modified?
+      @current_content != @parser.serialize(data)
+    end
+
+  private
+
+    # The path for loading the content
+    #
+    # @return [String] path to local configuration file if it exists; path to vendor one otherwise
+    def load_path
+      Yast::FileUtils.Exists(LOCAL_PATH) ? LOCAL_PATH : VENDOR_PATH
+    end
+
+    # The path for writing the content
+    #
+    # @return [String] path to local configuration file
+    def write_path
+      LOCAL_PATH
+    end
+
+    # Return the CFA::Matcher for given database name
+    #
+    # @param db_name [String] the database name, e.g., "passwd" or "hosts"
+    #
+    # @return [CFA::Matcher]
+    def database_matcher(db_name)
+      CFA::Matcher.new { |k, v| k == "database[]" && v.value == db_name }
+    end
+
+    # Find or create the DatabaseEntry for given database name
+    #
+    # @param db_name [String] the database name, e.g., "passwd" or "hosts"
+    #
+    # @return [DatabaseEntry]
+    def entry_for(db_name)
+      return databases[db_name] if databases.keys.include?(db_name)
+
+      database = CFA::AugeasTreeValue.new(CFA::AugeasTree.new, db_name)
+      data.add("database[]", database)
+      DatabaseEntry.new(database)
+    end
+
+    # Available databases
+    #
+    # @return [Hash{String => DatabaseEntry}]
+    def databases
+      raw_databases.each_with_object({}) do |database, collection|
+        name = database[:value].value.to_s
+        collection[name] = DatabaseEntry.new(database[:value])
+      end
+    end
+
+    # Available databases in internal structure
+    #
+    # @return [Array<Hash{Symbol => CFA::AugeasTreeValue, String, Symbol}>]
+    def raw_databases
+      data.select(CFA::Matcher.new(collection: "database"))
+    end
+
+    # Known collection keys
+    COLLECTION_KEYS = ["database", "service"].freeze
+    private_constant :COLLECTION_KEYS
+
+    # Fix collection names adding the missing "[]"
+    #
+    # When a collection has only one element, Augeas does not add "[]" suffix, reason why is
+    # necessary to fix at least those collections that are going to be hitted for either, read or
+    # write.
+    #
+    # @param tree [AugeasTree,AugeasTreeValue]
+    def fix_collection_names(tree)
+      tree.data.each do |entry|
+        entry[:key] << "[]" if COLLECTION_KEYS.include?(entry[:key])
+        fix_collection_names(entry[:value].tree) if entry[:value].is_a?(AugeasTreeValue)
+        fix_collection_names(entry[:value]) if entry[:value].is_a?(AugeasTree)
+      end
+    end
+
+    # Internal class for managing database entries
```

Comment:
:+1: that it operates on data itself and ot not try to make some fancy thing. Here we failed hard in past with ntp-client which contain a lot of buggy behavior.